### PR TITLE
touch: fix issue where directory is created instead of file

### LINKED ISSUE #5952: Rclone “touch” command fails since 1.57.0
#### The associated forum post URL from `https://forum.rclone.org`
https://forum.rclone.org/t/rclone-touch-command-fails-since-1-57-0/28890

#### What is the problem you are having with rclone?
Something changed with rclone 1.57.0 so that the rclone "touch" command is not usable for me anymore.
It all worked fine, and still does when I run rclone with 1.56.2, but it fails with newer versions.
I see that the --recursive option has been added to the "touch" command with 1.57.0, so maybe it is related to this addition?

Background: I use the "touch" command to test if the remote server is set up correctly. So I "touch" a new file that does not exist on the remote server and delete the file after success.

The errors I get with type FTP:
```
2022/01/26 12:47:06 ERROR : Attempt 1/1 failed with 1 errors and: failed to touch (create): update stor: 550 /files/ftp: Not a regular file
2022/01/26 12:47:06 Failed to touch: failed to touch (create): update stor: 550 /files/ftp: Not a regular file
```

The errors I get with type SFTP:
```
2022/01/26 13:01:26 ERROR : Attempt 1/1 failed with 1 errors and: failed to touch (create): Update Create failed: sftp: "Failure" (SSH_FX_FAILURE)
2022/01/26 13:01:26 Failed to touch: failed to touch (create): Update Create failed: sftp: "Failure" (SSH_FX_FAILURE)
```

#### What is your rclone version (output from `rclone version`)

```
rclone v1.57.0
- os/version: ubuntu 20.04 (64 bit)
- os/kernel: 5.4.0-94-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.2
- go/linking: static
- go/tags: none
```

I have also tested the latest Beta, but without success:

```
rclone v1.58.0-beta.5960.a667e03fc
- os/version: ubuntu 20.04 (64 bit)
- os/kernel: 5.4.0-94-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.17.6
- go/linking: static
- go/tags: none
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Ubuntu 20.04 - 64 bit

#### Which cloud storage system are you using? (e.g. Google Drive)
I have tested FTP, SFTP, Dropbox -> All report the same: "failed to touch"

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
With FTP:
```
rclone --config '/tmp/rclone-ftp.config' --retries 1 touch 'rclone-storage:/files/ftp'
```
-> Result: Nothing happens on the remote side. An rclone error message is shown (see above).

With SFTP:
```
rclone --config '/tmp/rclone-sftp.config' --retries 1 touch 'rclone-storage:/files/sftp'
```
-> Result: On the remote side, a new directory "sftp" is created in "/files/" (??? Why a directory ???). An rclone error message is shown (see above).

#### The rclone config contents with secrets removed.

```
[rclone-storage]
type = ftp
host = dev23.keymachine.de
user = test
pass = <secret-pass>
explicit_tls = true
```

```
[rclone-storage]
type = sftp
host = dev23.keymachine.de
user = test
pass = <secret-pass>
md5sum_command = none
sha1sum_command = none
```

#### A log from the command with the `-vv` flag

With FTP:
```
2022/01/26 12:47:28 DEBUG : rclone: Version "v1.57.0" starting with parameters ["rclone" "--config" "/tmp/rclone-ftp.config" "-vv" "--retries" "1" "touch" "rclone-storage:/files/ftp"]
2022/01/26 12:47:28 DEBUG : Creating backend with remote "rclone-storage:/files/ftp"
2022/01/26 12:47:28 DEBUG : Using config file from "/tmp/rclone-ftp.config"
2022/01/26 12:47:28 DEBUG : ftp://dev23.keymachine.de:21/files/ftp: Connecting to FTP server
2022/01/26 12:47:28 DEBUG : Touch time 2022-01-26 12:47:28.42150678 +0100 CET m=+0.361783329
2022/01/26 12:47:28 DEBUG : ftp://dev23.keymachine.de:21/files/ftp: Touching (creating)
2022/01/26 12:47:29 DEBUG : ftp://dev23.keymachine.de:21/files/ftp: Connecting to FTP server
2022/01/26 12:47:29 DEBUG : : Removed after failed upload: 550 /files/ftp: Not a regular file
2022/01/26 12:47:29 ERROR : Attempt 1/1 failed with 1 errors and: failed to touch (create): update stor: 550 /files/ftp: Not a regular file
2022/01/26 12:47:29 DEBUG : 2 go routines active
2022/01/26 12:47:29 Failed to touch: failed to touch (create): update stor: 550 /files/ftp: Not a regular file
```

With SFTP:
```
2022/01/26 13:04:07 DEBUG : rclone: Version "v1.57.0" starting with parameters ["rclone" "--config" "/tmp/rclone-sftp.config" "-vv" "--retries" "1" "touch" "rclone-storage:/files/sftp"]
2022/01/26 13:04:07 DEBUG : Creating backend with remote "rclone-storage:/files/sftp"
2022/01/26 13:04:07 DEBUG : Using config file from "/tmp/rclone-sftp.config"
2022/01/26 13:04:07 DEBUG : sftp://test@dev23.keymachine.de:22//files/sftp: New connection 62.141.56.104:51682->62.141.56.45:22 to "SSH-2.0-OpenSSH_8.4p1 Debian-5"
2022/01/26 13:04:07 DEBUG : Touch time 2022-01-26 13:04:07.852656441 +0100 CET m=+0.612075729
2022/01/26 13:04:07 DEBUG : sftp://test@dev23.keymachine.de:22//files/sftp: Touching (creating)
2022/01/26 13:04:07 ERROR : Attempt 1/1 failed with 1 errors and: failed to touch (create): Update Create failed: sftp: "Failure" (SSH_FX_FAILURE)
2022/01/26 13:04:07 DEBUG : 11 go routines active
2022/01/26 13:04:07 Failed to touch: failed to touch (create): Update Create failed: sftp: "Failure" (SSH_FX_FAILURE)
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

Thank you very much for your help!
Alex

### PR DESCRIPTION
**Pull request by albertony**
#### Was the change discussed in an issue or in the forum before?

Fixes #5952

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/ftp/ftp.go [hunk 1]
```diff
@@ -868,11 +868,15 @@ func (f *Fs) mkdir(ctx context.Context, abspath string) error {
...
 }
 
-// mkParentDir makes the parent of remote if necessary and any
-// directories above that
-func (f *Fs) mkParentDir(ctx context.Context, remote string) error {
-	parent := path.Dir(remote)
-	return f.mkdir(ctx, path.Join(f.root, parent))
+// mkParentDir makes the parent of the native path dirPath if
+// necessary and any directories above that
+func (f *Fs) mkParentDir(ctx context.Context, dirPath string) (err error) {
```
Comment (ncw) [inline backend/ftp/ftp.go:873]: Can we fix this without making `mkParentDir` take an absolute path. In my experience its best to keep the functions which take absolute paths down to a minimum.

File: backend/sftp/sftp.go [hunk 2]
```diff
@@ -972,11 +972,15 @@ func (f *Fs) PutStream(ctx context.Context, in io.Reader, src fs.ObjectInfo, opt
...
 }
 
-// mkParentDir makes the parent of remote if necessary and any
-// directories above that
-func (f *Fs) mkParentDir(ctx context.Context, remote string) error {
-	parent := path.Dir(remote)
-	return f.mkdir(ctx, path.Join(f.absRoot, parent))
+// mkParentDir makes the parent of the native path dirPath if
+// necessary and any directories above that
+func (f *Fs) mkParentDir(ctx context.Context, dirPath string) (err error) {
```
Comment (ncw) [inline backend/sftp/sftp.go:977]: Same here...

**Review by ncw — COMMENTED** [review]
Can you write a test in [fstests.go](https://github.com/rclone/rclone/blob/master/fstest/fstests/fstests.go) which fails for the ftp/sftp/dropbox backends? I think the underlying problem is that the `Put` method is failing to write a single file to the root - is that right? That should be easy enough to test for.

Then can you re-organize these patches so they are one per backend  to fix the test. Then this will never happen again as we'll test it in the integration tests :-) We'll probably find other backends with the same problem would be my guess.

File: backend/ftp/ftp.go (hunk 1 above)
Comment (albertony) [reply backend/ftp/ftp.go:873]: Do this internally in `Put` instead, you mean? I did something like that first, but then found that a `mkParentDirs` function exists in a few other backends already (webdav, jottacloud, mailru, yandex) - all of which had absolute path as argument.

**Comment by albertony** [comment]
I guess you meant more like this (pushed an update)? So the mkParentDir takes just the remote as before, but now joins it with root before finding and creating parents?

**Review by ncw — COMMENTED** [review]
This code is a lot neater :-) I'm not sure it is correct though!

After some time thinking about what exactly is going on I came up with this

## Before

```go
// remote = ""
// f.root = "file.txt" 
parent := path.Dir(remote)
// parent = "."
return f.mkdir(ctx, path.Join(f.root, parent))
// path.Join(f.root, parent) = "file.txt"
// => makes the "file.txt" directory
```

## After

```go
// remote = ""
// f.root = "file.txt" 
parent := path.Dir(path.Join(f.root, remote))
// path.Join(f.root, remote) = file.txt
// parent = "."
if parent == "." {
    parent = ""
}
// parent = ""
return f.mkdir(ctx, parent)
```

## Discussion

The problem stems from the behaviour of opening a backend pointing to a non-existent file let's say `remote:file.txt`

When the file doesn't exist, rclone assumes it is a directory and returns the backend pointing to that directory (`remote:file.txt`), with the root directory as yet uncreated.

When you write a file, the root directory gets created. All good so far.

However what `touch` seems to be doing is writing a file with an empty name using `f.Put("")` to the remote called (`remote:file.txt`).

This seems a bit odd to me and I'm not suprised it doesn't work on some backends! There certainly aren't integration tests for it. If we want it to work they we definitely need an integration test for it.

Wouldn't it be better for touch to parse the incoming path itself when --recursive isn't set using `fspath.Split`? Then it could be just creating the backend `remote:` and creating the file `file.txt` in it which doesn't seem problematic!

What do you think?

**Comment by albertony** [comment]
> However what `touch` seems to be doing is writing a file with an empty name using `f.Put("")` to the remote called (`remote:file.txt`).

Correct. And I agree it seems a bit strange, when you put it that way...

> Wouldn't it be better for touch to parse the incoming path itself when --recursive isn't set using `fspath.Split`? Then it could be just creating the backend `remote:` and creating the file `file.txt` in it which doesn't seem problematic!

Yes, probably! I will look into it. Thanks.

**Comment by albertony** [comment]
Yes, the NewFsDstFile is what was used before the recursive touch implementation.

But, this is only testing if --recursive is present or not, right? So, if I were to `rclone touch remote:temp.txt --recursive` for some strange reason, then on ftp and sftp it would still hit the same issue - strange (for users) error message and side-effect of creating a directory with name `temp.txt`..?

**Comment by ncw** [comment]
> Yes, the NewFsDstFile is what was used before the recursive touch implementation.
> 
> But, this is only testing if --recursive is present or not, right? So, if I were to `rclone touch remote:temp.txt --recursive` for some strange reason, then on ftp and sftp it would still hit the same issue - strange (for users) error message and side-effect of creating a directory with name `temp.txt`..?

Hmm, that is a tricky one.

I think the best solution is that we document that `--recursive` needs a directory.

I've written about 10 different answers to this and deleted them, but I decided to try it and the answer looks sensible and doesn't create unexpected directories.

```
$ rclone touch -vv --recursive --retries 1 :sftp,host=localhost,user=sftptest:doesnotexist11.txt
2022/01/27 17:17:49 NOTICE: sftp://sftptest@localhost:22/doesnotexist11.txt: Not touching non-existent file due to --recursive
```

```
$ rclone touch -vv --recursive --retries 1 :sftp,host=localhost,user=sftptest:normal-file.txt
2022/01/27 17:20:05 DEBUG : Touch time 2022-01-27 17:20:05.025510659 +0000 GMT m=+0.446746561
2022/01/27 17:20:05 DEBUG : sftp://sftptest@localhost:22/normal-file.txt: Touching "normal-file.txt"
```

Which looks sensible

**Comment by albertony** [comment]
> Not touching non-existent file due to --recursive

I see, it gets caught by that check! Thats good, then.

I'm testing your code change myself now, on top of clean master...

Found a different issue: When touching an existing directory without --recursive it touches the files in the parent. E.g. with:

```
.\rclone touch local_sftp:parent/existing_folder -vv
```

It logs:
> DEBUG : Touching files in directory non-recursively

And then "Touching" on all files in `parent`.

If appending a trailing slash:

```
.\rclone touch local_sftp:parent/existing_folder/ -vv
```

It does almost nothing, just exists with an (error??) message:

> 2022/01/27 18:59:42 DEBUG : rclone: Version "v1.58.0-DEV" starting with parameters [".\\rclone" "touch" "local_sftp:parent/existing_directory/" "-vv"]
> 2022/01/27 18:59:42 "local_sftp:parent/existing_directory/" is a directory

**Comment by albertony** [comment]
Pushed a new version - with your fix and the additional change I suggested above.

Unit test runs ok, integration test runs ok on sftp remote. Manual testing ok so far, but will test some more...

I removed (moved into a different branch) the original ones, they are not needed for touch - and I guess we do not want to include them then?

**Comment by albertony** [comment]
> Edit: Well, there is the mentioned is a directory case - looking into that, at least.

Thats a fatal error from the NewFsDstFile when there is no (potential) filename. E.g. path with ending slash. (I'm sure you know, but I didn't).

Maybe we have to split the path within touch after all? Pushed a version doing this. Need to "review" it myself, and test more - after a break.

**Comment by albertony** [comment]
@ncw: PTAL

The last version seems to work. It handles corner cases like:
- Root does not exist: `rclone touch remote:`
	- Skipped: Not touching empty directory
- Path does not exist and ends with slash: `rclone touch remote:notexists/`
	- Skipped: Not touching empty directory
- Path exist and ends with slash: `rclone touch remote:exists/`
	- Touching all files in the directory.
- Path does not exist: `rclone touch remote:touch.me`
	- if !recursive: Creates as file touch.me. If remote: points to a directory that does not exists it will be created too.
	- if recursive: Skipped: Not touching non-existent file due to --recursive

**Review by ncw — APPROVED** [review]
I think this looks good now - well done :-)

It was a bit of a subtle problem wasn't it!

**Comment by albertony** [comment]
> I think this looks good now - well done :-)

Great, thanks!

> It was a bit of a subtle problem wasn't it!

Yes, it was! I guess part of the reason was all the file vs folder + existing vs nonexisting + recursive vs nonrecursive combinations that was hard to fit perfectly in existing fs functions etc.

**Comment by albertony** [comment]
PS: This was merged into master before it was closed.
